ASAN_TRAP | WebCore::RenderObject::repaintRectangle

anil-hebbar opened the pull request:
#### 6e216e69281c511606d800818a1d83f7d888f5f2
<pre>
ASAN_TRAP | WebCore::RenderObject::repaintRectangle
<a href="https://bugs.webkit.org/show_bug.cgi?id=290013">https://bugs.webkit.org/show_bug.cgi?id=290013</a>
<a href="https://rdar.apple.com/146074591">rdar://146074591</a>

Reviewed by Alan Baradlay.

The issue was the the repaint rect was called in the middle of detach. In that state,
some of the memebers were detached and the others not, causing calls to detached
members. The fix was to call the repaint is called in the tree updater function.

* LayoutTests/editing/selection/selection-update-repaint-expected.txt: Added.
* LayoutTests/editing/selection/selection-update-repaint.html: Added.
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::resetRendererStateOnDetach):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::repaintAndMarkContainingBlockDirtyBeforeTearDown):

Canonical link: <a href="https://commits.webkit.org/293475@main">https://commits.webkit.org/293475@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5dece7e1e6656423ab48a2d7f74181da6865546d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98923 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18560 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104049 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49514 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100968 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18855 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27010 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75306 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32434 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101927 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14324 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89336 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55667 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14115 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7310 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48892 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84052 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7384 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106418 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18966 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84270 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26395 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83771 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21264 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28422 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6098 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19742 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25973 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31159 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25793 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29113 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27367 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -975,8 +975,8 @@ void RenderObject::propagateRepaintToParentWithOutlineAutoIfNeeded(const RenderL
             CheckedPtr enclosingMultiColumnFlow = previousMultiColumnSet->multiColumnFlow();
             CheckedPtr renderMultiColumnPlaceholder = enclosingMultiColumnFlow->findColumnSpannerPlaceholder(downcast<RenderBox>(renderer.get()));
-            ASSERT(renderMultiColumnPlaceholder);
-            renderer = WTFMove(renderMultiColumnPlaceholder);
```
It would be interesting to know what makes the placeholder disappear.

alanbaradlay commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -775,6 +776,11 @@ static std::optional<DidRepaintAndMarkContainingBlock> repaintAndMarkContainingB
+        // avoid problems of invalid pointers.
+        if (renderer.isSelectionBorder())
+            renderer.frame().selection().setNeedsSelectionUpdate();
+
```
I think this won't take care of RenderText type of renderers (this function seems to be operating on RenderElements while RenderTexts are cleaned up at tearDownTextRenderer).

anil-hebbar replied:
Updated the code to take care of both Text and Elements

rniwa commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -759,6 +760,15 @@ static std::optional<DidRepaintAndMarkContainingBlock> repaintAndMarkContainingB
         markContainingBlockDirty(*destroyRootRenderer);
     }
 
+    for (auto it = composedTreeDescendantsIterator.begin(), end = composedTreeDescendantsIterator.end(); it != end; ++it) {
```
I don't think we should add a new for loop.
We should just do this in the for loop below.

rniwa reviewed (changes requested):
We shouldn't be adding a new for loop like this.

rniwa commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -760,6 +761,13 @@ static std::optional<DidRepaintAndMarkContainingBlock> repaintAndMarkContainingB
 
     for (auto it = composedTreeDescendantsIterator.begin(), end = composedTreeDescendantsIterator.end(); it != end; ++it) {
+        if (it->renderer()) {
+            auto& renderer = *it->renderer();
```
We shouldn't be calling renderer() twice like this.
Just assign renderer to a local variable like this:
```
auto* renderer = it->renderer();
if (!renderer)
    continue;
```

rniwa commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -760,6 +761,13 @@ static std::optional<DidRepaintAndMarkContainingBlock> repaintAndMarkContainingB
+                renderer.frame().selection().setNeedsSelectionUpdate();
+        }
         auto* element = dynamicDowncast<Element>(*it);
         if (!element || !element->renderer())
```
This renderer() check is redundant.

rniwa commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -760,25 +761,35 @@ static std::optional<DidRepaintAndMarkContainingBlock> repaintAndMarkContainingB
+        if (!element)
             continue;
-        auto& renderer = *element->renderer();
+        auto& renderElement = *element->renderer();
```
Instead of getting the renderer out of the element again, we should just `dynamicDowncast<RenderElement>(renderer)`.

rniwa commented on LayoutTests/editing/selection/selection-update-repaint.html:
```diff
@@ -0,0 +1,59 @@
+<!-- url=(0014)about:internet -->
```
We need to simplify this test more.

rniwa replied:
Here's a reduced test case with reliable reproducibility:
```html
<!DOCTYPE html>
<html>
<head>
<style>
table { -webkit-column-count: 3; transform-origin: 81; column-span: all; }
.multicol { -webkit-column-count: 7; }
</style>
<script>
if (window.testRunner)
    testRunner.waitUntilDone();

function runTest() {
    getSelection().deleteFromDocument();
    setTimeout(() => {
        let count = location.search ? +location.search.substring(1) : 1;
        if (count >= 20) {
            document.open();
            document.write('This test passes if WebKit does not crash.<br>PASS.');
            document.close();
            if (window.testRunner)
                testRunner.notifyDone();
        } else
            location.href = '?' + (count + 1);
    }, 10);
}

function selectAll() {
    document.execCommand("selectAll", false);
}

</script>
</head>
<body onload=runTest()>
<div class="multicol" tabindex="1" autofocus>
    <table>
        <th>
            <table>a</table>
        </th>
        <tr>b</tr>
    </table>
</div>
<li class="multicol">
    <iframe allowfullscreen="true" srcdoc=")#&quot;~KVv8CRs" loopstart="1">c</iframe>
</li>
<details open="true" ontoggle="selectAll()">d</details>
</body>
```

anil-hebbar committed:
ASAN_TRAP | WebCore::RenderObject::repaintRectangle
https://bugs.webkit.org/show_bug.cgi?id=290013
rdar://146074591

Reviewed by Alan Baradlay.

The issue was the the repaint rect was called in the middle of detach. In that state,
some of the memebers were detached and the others not, causing calls to detached
members. The fix was to call the repaint is called in the tree updater function.

* LayoutTests/editing/selection/selection-update-repaint-expected.txt: Added.
* LayoutTests/editing/selection/selection-update-repaint.html: Added.
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
(WebCore::resetRendererStateOnDetach):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::repaintAndMarkContainingBlockDirtyBeforeTearDown):

Canonical link: https://commits.webkit.org/293475@main
